### PR TITLE
Support association joins from main query to be used in subqueries

### LINKED ISSUE #3334: Exception executing HQL query with uncorrelated left joins in subselect
### The base situation

We have a data model like this one
```
Entity ---0..* ChildEntity
       ---0..1 OtherEntity

ChildEntity ---1 GrandChildEntity
```

The associations to OtherEntity and GrandChildEntity are mapped as ```many-to-one```.

When executing the HQL query ...
```sql
SELECT ROOT 
FROM Entity AS ROOT 
WHERE
EXISTS 
    (FROM ELEMENTS(ROOT.Children) AS child
        LEFT JOIN child.Child AS grandChild
        LEFT JOIN ROOT.OtherEntity AS otherEntity
    WHERE
        grandChild.Name like 'G%'
        OR otherEntity.Name like 'G%'
    )
```

### The problem

... we get the following sql generated:
```sql
select entity0_.Id as id1_0_3_, entity0_.Name as name2_0_3_, entity0_.OtherEntity as otherentity3_0_3_
from Entity entity0_ 
where exists (
    select children1_.Id, grandchild2_.Id, otherentit3_.Id
    from ChildEntity children1_
        left outer join GrandChildEntity grandchild2_ on children1_.Child=grandchild2_.Id, 
        left outer join OtherEntity otherentit3_ on entity0_.OtherEntity=otherentit3_.Id 
    where entity0_.Id=children1_.Parent and (grandchild2_.Name like 'G%' or otherentit3_.Name like 'G%'))
```

This leads to an exception because the comma ```,``` before the last ```left outer join``` is illegal.

### Side note on why we don't use direct dot-navigations instead of explicit joins

Using implied joins on the :1-navigations like in the following query does not work because then, the joins are created as ```inner join``` or theta join so the condition is not evaluated correctly if there is not OtherEntity associated with Entity:
```sql
SELECT ROOT 
FROM Entity AS ROOT 
WHERE
    EXISTS 
        (FROM ELEMENTS(ROOT.Children) AS child
        WHERE
        child.Child.Name like 'G%'
        OR ROOT.OtherEntity.Name like 'A%'
)
```

I might start a separate discussion on Nullability and ```many-to-one``` because of that.

### PR DESCRIPTION
**Pull request by bahusoid**
Fixes #3334
Closes #3338
Closes #3335

### CODE REVIEW

File: src/NHibernate.Test/NHSpecificTest/GH1228/Fixture.cs [hunk 1]
```diff
@@ -7,126 +7,95 @@ public class Fixture : BugTestCase
 		[Test]
 		public void TestOk()
 		{
-			using (ISession s = OpenSession())
+			using var s = OpenSession();
 			{
```
Comment (fredericDelaporte) [inline src/NHibernate.Test/NHSpecificTest/GH1228/Fixture.cs:11]: I find it confusing to use a block code after an "instruction line" using.

```suggestion
			using var s = OpenSession();
```

File: src/NHibernate.Test/NHSpecificTest/GH1228/Fixture.cs [hunk 2]
```diff
@@ -7,126 +7,95 @@ public class Fixture : BugTestCase
...
-						{
-							var queryThatWorks = s.CreateQuery(@"
+				var queryThatWorks = s.CreateQuery(
+					@"
 							SELECT ROOT FROM NHibernate.Test.NHSpecificTest.GH1228.Sheet AS ROOT 
 							WHERE (EXISTS (FROM NHibernate.Test.NHSpecificTest.GH1228.Shelf AS inv  
 										   , ROOT.Folder AS ROOT_Folder 
 										   WHERE ROOT_Folder.Shelf = inv AND inv.Id = 1
 								  )	   ) 
 								  AND ROOT.Name = 'SomeName'");
```
Comment (fredericDelaporte) [inline src/NHibernate.Test/NHSpecificTest/GH1228/Fixture.cs:19]: Resulting indentation adjustment:

```suggestion
			var queryThatWorks = s.CreateQuery(
				@"
				SELECT ROOT FROM NHibernate.Test.NHSpecificTest.GH1228.Sheet AS ROOT
					WHERE (EXISTS (FROM NHibernate.Test.NHSpecificTest.GH1228.Shelf AS inv
									, ROOT.Folder AS ROOT_Folder
								WHERE ROOT_Folder.Shelf = inv AND inv.Id = 1
							))
						AND ROOT.Name = 'SomeName'");
```

File: src/NHibernate.Test/NHSpecificTest/GH1228/Fixture.cs [hunk 3]
```diff
@@ -7,126 +7,95 @@ public class Fixture : BugTestCase
...
-					finally
-					{
-						s.Delete("from Sheet");
-						s.Delete("from Folder");
-						s.Delete("from Shelf");
-						t.Commit();
-					}
-				}
+				queryThatWorks.List();
 			}
```
Comment (fredericDelaporte) [inline src/NHibernate.Test/NHSpecificTest/GH1228/Fixture.cs:32]: End of resulting changes:

```suggestion
			queryThatWorks.List();
```

File: src/NHibernate.Test/NHSpecificTest/GH1228/Fixture.cs [hunk 4]
```diff
@@ -7,126 +7,95 @@ public class Fixture : BugTestCase
...
+				// The only assertion here is that the generated SQL is valid and can be executed.
+				// Right now, the generated SQL is missing the JOIN inside the subselect (EXISTS) to Folder.
+				var queryThatCreatesWrongSQL = s.CreateQuery(
+					@"
 							SELECT ROOT FROM NHibernate.Test.NHSpecificTest.GH1228.Shelf AS ROOT 
 							WHERE (EXISTS (FROM NHibernate.Test.NHSpecificTest.GH1228.Sheet AS sheet
 										   JOIN ROOT.Folders AS ROOT_Folder 
 										   WHERE ROOT_Folder = sheet.Folder AND sheet.Name = 'SomeName'
 								  )	   ) 
 								  AND ROOT.Id = 1");
```
Comment (fredericDelaporte) [inline src/NHibernate.Test/NHSpecificTest/GH1228/Fixture.cs:63]:
```suggestion
			queryThatCreatesWrongSQL.List();

			// The only assertion here is that the generated SQL is valid and can be executed.
			// With the bug, the generated SQL is missing the JOIN inside the subselect (EXISTS) to Folder.
			queryThatCreatesWrongSQL = s.CreateQuery(
				@"
				SELECT ROOT FROM NHibernate.Test.NHSpecificTest.GH1228.Shelf AS ROOT
					WHERE (EXISTS (FROM NHibernate.Test.NHSpecificTest.GH1228.Sheet AS sheet
								JOIN ROOT.Folders AS ROOT_Folder
								WHERE ROOT_Folder = sheet.Folder AND sheet.Name = 'SomeName'
							))
						AND ROOT.Id = 1");
```

File: src/NHibernate.Test/NHSpecificTest/GH1228/Fixture.cs [hunk 5]
```diff
@@ -7,126 +7,95 @@ public class Fixture : BugTestCase
...
+				// The only assertion here is that the generated SQL is valid and can be executed.
+				// Right now, the generated SQL is missing the JOIN inside the subselect (EXISTS) to Folder.
+				var queryThatCreatesWrongSQL = s.CreateQuery(
+					@"
 							SELECT ROOT FROM NHibernate.Test.NHSpecificTest.GH1228.Shelf AS ROOT 
 							WHERE (EXISTS (FROM NHibernate.Test.NHSpecificTest.GH1228.Sheet AS sheet
 										   JOIN sheet.Folder AS folder
 										   WHERE folder.Shelf = ROOT AND sheet.Name = 'SomeName'
 								  )	   ) 
 								  AND ROOT.Id = 1");
```
Comment (fredericDelaporte) [inline src/NHibernate.Test/NHSpecificTest/GH1228/Fixture.cs:84]:
```suggestion
		public void TestOtherAnsiJoinOnAssociationInSubQuery()
		{
			using var s = OpenSession();

			// The only assertion here is that the generated SQL is valid and can be executed.
			// With the bug, the generated SQL is missing the JOIN inside the subselect (EXISTS) to Folder.
			var queryThatCreatesWrongSQL = s.CreateQuery(
				@"
				SELECT ROOT FROM NHibernate.Test.NHSpecificTest.GH1228.Shelf AS ROOT
					WHERE (EXISTS (FROM NHibernate.Test.NHSpecificTest.GH1228.Sheet AS sheet
								JOIN sheet.Folder AS folder
								WHERE folder.Shelf = ROOT AND sheet.Name = 'SomeName'
							))
						AND ROOT.Id = 1");
```

File: src/NHibernate.Test/NHSpecificTest/GH1228/Fixture.cs [hunk 6]
```diff
@@ -7,126 +7,95 @@ public class Fixture : BugTestCase
...
+			}
+			{
+				var queryThatCreatesWrongSQL = s.CreateQuery(
+					@"
 							SELECT ROOT FROM NHibernate.Test.NHSpecificTest.GH1228.Sheet AS ROOT 
 							WHERE (EXISTS (FROM NHibernate.Test.NHSpecificTest.GH1228.Shelf AS inv  
 										   JOIN inv.Folders AS folder
 										   WHERE folder = ROOT.Folder AND inv.Id = 1
 								  )	   ) 
 								  AND ROOT.Name = 'SomeName'");
```
Comment (fredericDelaporte) [inline src/NHibernate.Test/NHSpecificTest/GH1228/Fixture.cs:97]:
```suggestion
			queryThatCreatesWrongSQL.List();
			
			// The only assertion here is that the generated SQL is valid and can be executed.
			// With the bug, the generated SQL is missing the JOIN inside the subselect (EXISTS) to Folder.
			queryThatCreatesWrongSQL = s.CreateQuery(
				@"
				SELECT ROOT FROM NHibernate.Test.NHSpecificTest.GH1228.Sheet AS ROOT
					WHERE (EXISTS (FROM NHibernate.Test.NHSpecificTest.GH1228.Shelf AS inv
								JOIN inv.Folders AS folder
								WHERE folder = ROOT.Folder AND inv.Id = 1
							))
						AND ROOT.Name = 'SomeName'");
```

File: src/NHibernate.Test/NHSpecificTest/GH1228/Fixture.cs [hunk 7]
```diff
@@ -7,126 +7,95 @@ public class Fixture : BugTestCase
...
-					}
-					finally {
-						s.Delete("from Sheet");
-						s.Delete("from Folder");
-						s.Delete("from Shelf");
-						t.Commit();
-					}
-				}
+				queryThatCreatesWrongSQL.List();
 			}
```
Comment (fredericDelaporte) [inline src/NHibernate.Test/NHSpecificTest/GH1228/Fixture.cs:99]:
```suggestion
			queryThatCreatesWrongSQL.List();
```

File: src/NHibernate.Test/NHSpecificTest/GH3334/Entity.cs [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+namespace NHibernate.Test.NHSpecificTest.GH3334
+{
+	public class Entity
+	{
+		public virtual int Id { get; set; }
+ 		public virtual string Name { get; set; }
+        public virtual ISet<ChildEntity> Children { get; set; } = new HashSet<ChildEntity>();
+        public virtual OtherEntity OtherEntity { get; set; }
+	}
```
Comment (fredericDelaporte) [inline src/NHibernate.Test/NHSpecificTest/GH3334/Entity.cs:11]: Mixed tab and spaces.
```suggestion
	{
		public virtual int Id { get; set; }
 		public virtual string Name { get; set; }
		public virtual ISet<ChildEntity> Children { get; set; } = new HashSet<ChildEntity>();
		public virtual OtherEntity OtherEntity { get; set; }
	}
```

**Comment by bahusoid** [comment]
Adding to single batch all those changes are quite annoying. Next time please push directly your formatting suggestions.

**Review by fredericDelaporte — DISMISSED** [review]

File: src/NHibernate.Test/NHSpecificTest/GH1228/Fixture.cs [hunk 9]
```diff
@@ -5,129 +5,90 @@ namespace NHibernate.Test.NHSpecificTest.GH1228
 	public class Fixture : BugTestCase
```
Comment (fredericDelaporte) [inline src/NHibernate.Test/NHSpecificTest/GH1228/Fixture.cs:5]: No async counterpart due to missing `[TestFixture]` attribute.

```suggestion
	[TestFixture]
	public class Fixture : BugTestCase
```

**Review by fredericDelaporte — DISMISSED** [review]

**Review by fredericDelaporte — DISMISSED** [review]

**Review by fredericDelaporte — APPROVED** [review]

**Pull request merged**